service/storage_service.cc

```diff
@@ -3632,18 +3646,26 @@ future<> storage_service::handle_state_normal(inet_address endpoint, gms::permit
 
     auto tmlock = std::make_unique<token_metadata_lock>(co_await get_token_metadata_lock());
     auto tmptr = co_await get_mutable_token_metadata_ptr();
-    if (tmptr->is_normal_token_owner(endpoint)) {
-        slogger.info("Node {} state jump to normal", endpoint);
-    }
     std::unordered_set<inet_address> endpoints_to_remove;
 
     auto do_remove_node = [&] (gms::inet_address node) {
-        tmptr->remove_endpoint(node);
+        // this lambda is called in three cases:
+        // 1. old endpoint for the given host_id is ours, we remove the new endpoint;
+        // 2. new endpoint for the given host_id has bigger generation, we remove the old endpoint;
+        // 3. old endpoint for the given host_id has bigger generation, we remove the new endpoint.
+        // In all of these cases host_id is retained, only the IP addresses are changed.
+        // We don't need to call remove_endpoint on tmptr, since it will be called
+        // indirectly through the chain endpoints_to_remove->storage_service::remove_endpoint ->
+        // _gossiper.remove_endpoint -> storage_service::on_remove.
```

Comment:
Pretty complicated reasoning, depending on such deeply nested chain.
Is there a reason for not just keeping the `remove_endpoint` call as it was?
Since you explained only why we don't *need* it, but no mention of why we shouldn't do it.
(What was the reason it was present in the first place?)

Reply:
The entire `handle_state_normal` operates with IPs for now. We need IPs to remove the nodes from `system.peers` at least, which is what `storage_service::remove_endpoint` does. It's mentally simpler not to mix IPs and `host_id`-s. The comment with the call chain just explains why removing from tm by `host_id` is redundant here. This is the piece of knowledge I discovered by digging into all of this mess and I definitely don't want to loose it :)